relay-server/src/actors/global_config.rs

```diff
@@ -0,0 +1,143 @@
+        Self {
+            enabled,
+            sender,
+            upstream,
+        }
+    }
+
+    fn handle_message(&self, message: GlobalConfiguration) {
+        match message {
+            GlobalConfiguration::Get(sender) => {
+                sender.send(self.sender.borrow().clone());
+            }
+            GlobalConfiguration::Subscribe(sender) => {
+                sender.send(self.sender.subscribe());
+            }
+        }
+    }
+
+    /// Requests a new global config from upstream, sends it back to spawn handler to update the watch,
+    /// as we don't have mutable access to `self` in an async block.
+    fn update_global_config(&self, global_tx: UnboundedSender<Arc<GlobalConfig>>) {
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
+
+        tokio::spawn(async move {
+            let query = GetProjectStates {
+                public_keys: vec![],
+                full_config: false,
+                no_cache: false,
+                global: true,
+            };
+
+            match upstream_relay.send(SendQuery(query)).await {
+                Ok(Ok(response)) => match response.global {
+                    Some(global_config) => {
+                        let global_config = Arc::new(global_config);
+                        if let Err(e) = global_tx.send(global_config) {
+                            relay_log::error!("failed to send global config back to global configuration spawn handler: {}", e);
+                        }
+                    }
+                    None => relay_log::error!("global config is missing in upstream response"),
+                },
+                Ok(Err(e)) => {
+                    relay_log::error!("failed to fetch global config request: {}", e);
+                }
+                Err(e) => {
+                    relay_log::error!("failed to send global config request: {}", e);
+                }
+            };
+        });
+    }
+}
+
+impl Service for GlobalConfigurationService {
+    type Interface = GlobalConfiguration;
+
+    fn spawn_handler(self, mut rx: relay_system::Receiver<Self::Interface>) {
+        tokio::spawn(async move {
+            relay_log::info!("global configuration service starting");
+
+            let mut ticker = tokio::time::interval(Duration::from_secs(10));
+            // Channel for sending new global configs from upstream to the spawn loop, so that we may
+            // update the tokio::watch.
+            let (global_tx, mut global_rx) = mpsc::unbounded_channel();
+
+            loop {
+                tokio::select! {
+                    biased;
+                    Some(global_config) = global_rx.recv() => {
+                        if let Err(e) = self.sender.send(global_config) {
+                            relay_log::error!("failed to update global config watch: {}", e);
+                        };
+                    },
+                    _ = ticker.tick(), if self.enabled =>  {
```

Comment:
its not enough to just check the upstream authentication, because if it's not in managed mode, it will say it's authenticated even though it's not, so that's why i have an additional `enabled` flag which for now just checks if relay credentials are present. reading through the code, it seems credentials do not change during normal operation.